klippy/extras/sht4x.py

```diff
@@ -0,0 +1,262 @@
+import logging
+from . import bus
+import threading
+
+# SHT4x I2C Commands
+DEFAULT_ADDR = 0x44
+RESET = 0x94
+REQUEST_CHIPID = 0x89
+
+# Measurement commands
+HIGH_PRECISION_MODE = 0xFD    # ~8.3ms measurement time
+MEDIUM_PRECISION_MODE = 0xF6  # ~4.5ms measurement time
+LOW_PRECISION_MODE = 0xE0     # ~1.6ms measurement time
+
+# Timing constants from datasheet
+MEASUREMENT_TIMES = {
+    HIGH_PRECISION_MODE: 0.0083,   # 8.3ms max according to datasheet
+    MEDIUM_PRECISION_MODE: 0.0045, # 4.5ms max according to datasheet
+    LOW_PRECISION_MODE: 0.0016     # 1.6ms max according to datasheet
+}
+
+SOFT_RESET_TIME = 0.001  # 1ms max according to datasheet
+REPORT_TIME = 0.8        # How often to report readings
+
+
+class SHT4X:
+    def __init__(self, config):
+        self.printer = config.get_printer()
+        self.name = config.get_name().split()[-1]
+        self.reactor = self.printer.get_reactor()
+
+        # I2C setup
+        valid_addresses = [0x44, 0x45, 0x46]  # From datasheet
+        addr = config.getint('i2c_address', DEFAULT_ADDR)
+        if addr not in valid_addresses:
+            addr_list = ', '.join(['0x%x' % a for a in valid_addresses])
+            raise config.error("Invalid I2C address 0x%x. "
+                              "Valid addresses are: %s" % (addr, addr_list))
+
+        speed = config.getint('i2c_speed', 100000)
+        self.i2c = bus.MCU_I2C_from_config(
+            config, default_addr=addr, default_speed=speed)
+        self.mcu = self.i2c.get_mcu()
+
+        # Precision setup
+        precision = config.get('precision', 'high').lower()
+        precision_map = {
+            'high': HIGH_PRECISION_MODE,
+            'medium': MEDIUM_PRECISION_MODE,
+            'low': LOW_PRECISION_MODE
+        }
+        if precision not in precision_map:
+            raise config.error("Invalid precision value '%s'. "
+                              "Valid options are: high, medium, low"
+                              % precision)
+        self.precision_mode = precision_map[precision]
+
+        # Core sensor state
+        self.temp = 0
+        self.humidity = 0
+        self._error_count = 0
+        self.sensor_ready = False
+        self.measurement_lock = threading.Lock()
```

Comment:
AFAIU, you do not need locking here.
There are no threads.
The whole code runs inside the single-threaded reactor.
sample_sensor() - could not run, until you ask it to (at the end of connect), and until you have returned control to the outer loop, by `return` or `pause()`.